documentation/library_guide/parallel_api/pass_data_algorithms.rst

```diff
@@ -81,47 +76,66 @@ the buffer were created for the same queue. For example:
+    // each element of vec equals to 42
     return 0;
   }
 
-When using device USM, such as allocated by ``malloc_device``, manually copy data to this memory
-before calling oneDPL algorithms, and copy it back once the algorithms have finished execution.
-
-Use Host-Side std::vector
------------------------------
+When using iterators to host allocated data, a temporary SYCL buffer is created, and the data
+is copied to this buffer. After processing on a device is complete, the modified data is copied
+from the temporary buffer back to the host container. While convenient, using host allocated
+data can lead to unintended copying between host and device. We recommend working with SYCL buffers
+or USM memory to reduce data copying between the host and device. 
 
-|onedpl_long| parallel algorithms can be called with ordinary (host-side) iterators, as seen in the
-example below.
-In this case, a temporary SYCL buffer is created, and the data is copied to this buffer.
-After processing on a device is complete, the modified data is copied from the temporary buffer back
-to the host container.
-For example:
+You can also use ``std::vector`` with a USM allocator, as shown in the following example:
 
 .. code:: cpp
 
   #include <oneapi/dpl/execution>
   #include <oneapi/dpl/algorithm>
-  #include <vector>
+  #include <sycl/sycl.hpp>
   int main(){
-    std::vector<int> vec( 1000 );
-    std::fill(oneapi::dpl::execution::dpcpp_default, vec.begin(), vec.end(), 42);
-    // each element of vec equals to 42
+    const int n = 1000;
+    auto policy = oneapi::dpl::execution::dpcpp_default;
+    sycl::usm_allocator<int, sycl::usm::alloc::shared> alloc(policy.queue());
+    std::vector<int, decltype(alloc)> vec(n, alloc);
+
+    // Recommended to use USM pointers:
+    std::fill(policy, vec.data(), vec.data() + vec.size(), 42);
+
+    // Iterators for USM allocators might require extra copying
+    // std::fill(policy, vec.begin(), vec.end(), 42);
+
     return 0;
   }
 
-Working with SYCL buffers is recommended to reduce data copying between the host and device.
\ No newline at end of file
+Make sure that the execution policy and the USM-allocated memory were created for the same queue.
+When using USM allocators with ``std::vector``, it is recommended to use USM pointers as shown in
+the example above, rather than with iterators to ``std::vector``. In some implementations of the C++
+Standard Library, it is not possible to detect that iterators are pointing to USM-allocated data.
+In those cases, data will be treated as if it were host-side data, and will require extra copying
+between host and device. |onedpl_short| will avoid copies whenever it is possible with the C++
+Standard Library implementation.  However, to guarantee no unintended copying, use
+``std::vector::data()`` in combination with ``std::vector::size()`` when using USM-allocated data
+to retrieve USM pointers for the sequence as shown.
```

Comment:
Hopefully the following makes all the same points while being somewhat more concise.

```suggestion
For ``std::vector`` with a USM allocator we recommend to use ``std::vector::data()`` in
combination with ``std::vector::size()`` as shown in the example above, rather than iterators to
``std::vector``. That is because for some implementations of the C++ Standard Library it might not
be possible for |onedpl_short| to detect that iterators are pointing to USM-allocated data. In that
case the data will be treated as if it were host-allocated, with an extra copy made to a SYCL buffer.
Retrieving USM pointers from ``std::vector`` as shown guarantees no unintended copying.
```

Reply:
Yes, this gets to the point, and doesn't bog the user down with unnecessary details.  Thanks, suggestion taken.